osu.Game/Screens/OnlinePlay/Multiplayer/Match/MultiplayerUserModSelectOverlay.cs

```diff
@@ -0,0 +1,117 @@
+{
+    public partial class MultiplayerUserModSelectOverlay : UserModSelectOverlay
+    {
+        [Resolved]
+        private MultiplayerClient client { get; set; } = null!;
+
+        [Resolved]
+        private RulesetStore rulesets { get; set; } = null!;
+
+        private ModSettingChangeTracker? modSettingChangeTracker;
+        private ScheduledDelegate? debouncedModSettingsUpdate;
+
+        public MultiplayerUserModSelectOverlay()
+            : base(OverlayColourScheme.Plum)
+        {
+        }
+
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+
+            client.RoomUpdated += onRoomUpdated;
+            SelectedMods.BindValueChanged(onSelectedModsChanged);
+
+            updateValidMods();
+        }
+
+        private void onRoomUpdated() => Scheduler.AddOnce(updateValidMods);
+
+        private void onSelectedModsChanged(ValueChangedEvent<IReadOnlyList<Mod>> mods)
+        {
+            modSettingChangeTracker?.Dispose();
+
+            if (client.Room == null)
+                return;
+
+            client.ChangeUserMods(mods.NewValue).FireAndForget();
+
+            modSettingChangeTracker = new ModSettingChangeTracker(mods.NewValue);
+            modSettingChangeTracker.SettingChanged += _ =>
+            {
+                // Debounce changes to mod settings so as to not thrash the network.
+                debouncedModSettingsUpdate?.Cancel();
+                debouncedModSettingsUpdate = Scheduler.AddDelayed(() =>
+                {
+                    if (client.Room == null)
+                        return;
+
+                    client.ChangeUserMods(SelectedMods.Value).FireAndForget();
+                }, 500);
+            };
+        }
+
+        private void updateValidMods()
+        {
+            if (client.Room == null || client.LocalUser == null)
+                return;
+
+            MultiplayerPlaylistItem currentItem = client.Room.CurrentPlaylistItem;
+            Ruleset ruleset = rulesets.GetRuleset(client.LocalUser.RulesetId ?? currentItem.RulesetID)!.CreateInstance();
+            Mod[] allowedMods = currentItem.Freestyle
+                ? ruleset.AllMods.OfType<Mod>().Where(m => ModUtils.IsValidFreeModForMatchType(m, client.Room.Settings.MatchType)).ToArray()
+                : currentItem.AllowedMods.Select(m => m.ToMod(ruleset)).ToArray();
+
+            // Update the mod panels to reflect the ones which are valid for selection.
+            IsValidMod = allowedMods.Length > 0
+                ? m => allowedMods.Any(a => a.GetType() == m.GetType())
+                : _ => false;
+
+            // Remove any mods that are no longer allowed.
+            Mod[] newUserMods = SelectedMods.Value.Where(m => allowedMods.Any(a => m.GetType() == a.GetType())).ToArray();
+            if (!newUserMods.SequenceEqual(SelectedMods.Value))
+                SelectedMods.Value = newUserMods;
+
+            ActiveMods.Value = ComputeActiveMods();
```

Comment:
What's the reason for unconditionally calling this here? It's already supposed to get called on `SelectedMods` changes:

https://github.com/ppy/osu/blob/596b7038ba468f8386c0b5c2a069ba2338542bfa/osu.Game/Overlays/Mods/ModSelectOverlay.cs#L232-L238

Reply:
Added deduping and documented the behaviour in https://github.com/ppy/osu/pull/32649/commits/aa58fa58cb6f2d5a3679a60e5f8fb5f44fd7a8e0